File: generar_hash.py
# ============================================
# IPUC LA FONDA - Generador de Hash de Contraseña
# v2.1.0
# ============================================
# Este script genera el hash de una contraseña usando la misma
# SECRET_KEY que se usa en app.py.
# 
# INSTRUCCIONES:
# 1. Copia la SECRET_KEY de tu archivo app.py
# 2. Pégala en la variable SECRET_KEY de este script
# 3. Ejecuta: python generar_hash.py
# 4. Copia el hash generado y pégalo en data/administradores.json
# ============================================

import hashlib
import secrets
import sys


def generar_hash_contraseña():
    """
    Genera el hash de una contraseña usando SHA-256 + salt.
    """
    print("=" * 60)
    print("🔐 IPUC LA FONDA - Generador de Hash de Contraseña")
    print("=" * 60)
    print("")
    
    # ============================================
    # CONFIGURACIÓN - MODIFICA ESTOS VALORES
    # ============================================
    
    # ⚠️  IMPORTANTE: Copia aquí la SECRET_KEY de tu archivo app.py
    # La SECRET_KEY se encuentra al inicio de app.py:
    #   SECRET_KEY = secrets.token_hex(32)
    # Debes reemplazar esa línea por el valor fijo que se muestra abajo
    SECRET_KEY = "TU_SECRET_KEY_AQUI"  # <-- CAMBIA ESTO
    
    # Contraseña a hashear
    password = "S3QyFkrCkNdRQk4"  # <-- CAMBIA ESTO SI QUIERES OTRA CONTRASEÑA
    
    # ============================================
    # VALIDACIÓN
    # ============================================
    if SECRET_KEY == "TU_SECRET_KEY_AQUI":
        print("⚠️  ╔══════════════════════════════════════════════════════════╗")
        print("⚠️  ║  ADVERTENCIA                                            ║")
        print("⚠️  ║  No has configurado la SECRET_KEY.                      ║")
        print("⚠️  ║                                                          ║")
        print("⚠️  ║  Debes copiar la SECRET_KEY de tu archivo app.py        ║")
        print("⚠️  ║  y pegarla en la variable SECRET_KEY de este script.    ║")
        print("⚠️  ╚══════════════════════════════════════════════════════════╝")
        print("")
        print("📝 Pasos a seguir:")
        print("   1. Abre tu archivo app.py")
        print("   2. Al iniciar el servidor, busca la línea que dice:")
        print("      'Tu SECRET_KEY temporal: xxxxxxxxxxxxxxxx'")
        print("      O busca: SECRET_KEY = secrets.token_hex(32)")
        print("   3. Copia esa SECRET_KEY")
        print("   4. Pégala en este script y vuelve a ejecutarlo")
        print("")
        
        # Opción: generar una nueva SECRET_KEY
        print("💡 ¿Quieres generar una nueva SECRET_KEY?")
        respuesta = input("   (s/n): ").strip().lower()
        
        if respuesta in ['s', 'si', 'sí', 'y', 'yes']:
            SECRET_KEY = secrets.token_hex(32)
            print("")
            print("✅ Nueva SECRET_KEY generada:")
            print(f"   {SECRET_KEY}")
            print("")
            print("📝 Ahora debes actualizar app.py:")
            print("   Busca: SECRET_KEY = secrets.token_hex(32)")
            print("   Cambia por: SECRET_KEY = \"TU_NUEVA_SECRET_KEY\"")
            print("")
        else:
            print("")
            print("❌ No se puede continuar sin una SECRET_KEY válida.")
            print("   Ejecuta el script de nuevo cuando tengas la SECRET_KEY.")
            return
    
    # ============================================
    # GENERAR HASH
    # ============================================
    def hash_password(password, secret_key):
        """Hashea la contraseña con SHA-256 + salt"""
        salt = secret_key[:16]
        return hashlib.sha256(f"{password}{salt}".encode()).hexdigest()
    
    print("=" * 60)
    print("🔒 GENERANDO HASH DE CONTRASEÑA")
    print("=" * 60)
    print("")
    print(f"   SECRET_KEY:  {SECRET_KEY[:20]}... (truncado por seguridad)")
    print(f"   Contraseña:  {'*' * len(password)}")
    print("")
    
    hash_resultado = hash_password(password, SECRET_KEY)
    
    print("✅ Hash generado exitosamente:")
    print("")
    print("=" * 60)
    print(f"   {hash_resultado}")
    print("=" * 60)
    print("")
    print("📝 INSTRUCCIONES DE USO:")
    print("")
    print("   Opción 1: Usar en data/administradores.json")
    print("   Abre el archivo data/administradores.json y pega el hash")
    print("   en el campo 'password' del administrador.")
    print("")
    print("   Opción 2: Usar en app.py")
    print("   1. Abre app.py")
    print("   2. Busca: SECRET_KEY = secrets.token_hex(32)")
    print("   3. Cambia por: SECRET_KEY = \"TU_SECRET_KEY\"")
    print("   4. Reinicia el servidor Flask")
    print("")
    print("⚠️  IMPORTANTE:")
    print("   • La SECRET_KEY debe ser LA MISMA en app.py y en este script")
    print("   • Si cambias la SECRET_KEY, TODAS las contraseñas anteriores")
    print("     dejarán de funcionar")
    print("   • Guarda la SECRET_KEY en un lugar seguro")
    print("=" * 60)
    print("")
    
    # Preguntar si quiere guardar en archivo
    guardar = input("¿Quieres guardar el hash en un archivo hash.txt? (s/n): ").strip().lower()
    if guardar in ['s', 'si', 'sí', 'y', 'yes']:
        with open('hash.txt', 'w', encoding='utf-8') as f:
            f.write("=" * 60 + "\n")
            f.write("IPUC LA FONDA - Hash de Contraseña\n")
            f.write("=" * 60 + "\n")
            f.write(f"SECRET_KEY: {SECRET_KEY}\n")
            f.write(f"Contraseña: {password}\n")
            f.write(f"Hash: {hash_resultado}\n")
            f.write("=" * 60 + "\n")
            f.write("\n⚠️  Guarda este archivo en un lugar seguro\n")
        print("✅ Hash guardado en hash.txt")
    
    return hash_resultado


# ============================================
# EJECUCIÓN PRINCIPAL
# ============================================
if __name__ == "__main__":
    try:
        hash_generado = generar_hash_contraseña()
        
        if hash_generado:
            print("")
            print("🎉 ¡Hash generado correctamente!")
            print("   Ya puedes usar este hash en tu archivo administradores.json")
        else:
            print("")
            print("ℹ️  No se generó ningún hash.")
            
    except KeyboardInterrupt:
        print("\n\n⏹️  Operación cancelada por el usuario")
    except Exception as e:
        print(f"\n\n❌ Error inesperado: {str(e)}")
    
    print("")
    input("Presiona Enter para salir...")

File: generar_hash.js
// ============================================
// IPUC LA FONDA - Generador de Hash v5.0
// Genera hash de contraseña para usar en la app
// "Donde el Espíritu Santo se mueve"
// ============================================

function generarHash(password, salt = 'ipuc5_salt_2026') {
    let hash = 0;
    const str = password + salt;
    for (let i = 0; i < str.length; i++) {
        const char = str.charCodeAt(i);
        hash = ((hash << 5) - hash) + char;
        hash |= 0;
    }
    return Math.abs(hash).toString(16);
}

// ============================================
// INTERFAZ DE CONSOLA
// ============================================
function generarHashConsola() {
    console.log('='.repeat(60));
    console.log('🔐 IPUC LA FONDA v5.0 - Generador de Hash de Contraseña');
    console.log('='.repeat(60));
    console.log('');
    
    // Contraseña por defecto
    const password = "S3QyFkrCkNdRQk4";
    const salt = 'ipuc5_salt_2026';
    
    console.log('🔒 GENERANDO HASH DE CONTRASEÑA');
    console.log('-'.repeat(60));
    console.log(`   Salt:        ${salt}`);
    console.log(`   Contraseña:  ${'*'.repeat(password.length)}`);
    console.log('');
    
    const hashResultado = generarHash(password, salt);
    
    console.log('✅ Hash generado exitosamente:');
    console.log('');
    console.log('='.repeat(60));
    console.log(`   ${hashResultado}`);
    console.log('='.repeat(60));
    console.log('');
    console.log('📝 INSTRUCCIONES DE USO:');
    console.log('');
    console.log('   Opción 1: Usar en crear-admin.html');
    console.log('   El hash se genera automáticamente al crear el admin');
    console.log('');
    console.log('   Opción 2: Usar en data/administradores.json');
    console.log('   Pega el hash en el campo "password" del administrador');
    console.log('');
    console.log('   Opción 3: Usar directamente en la consola');
    console.log('   Ejecuta: crearAdminLocal("Nombre","Apellidos","correo","usuario","password")');
    console.log('');
    console.log('⚠️  IMPORTANTE:');
    console.log('   • El salt debe ser el mismo en database.js');
    console.log('   • Si cambias el salt, todas las contraseñas anteriores');
    console.log('     dejarán de funcionar');
    console.log('='.repeat(60));
    
    return hashResultado;
}

// ============================================
// FUNCIÓN PARA GUARDAR HASH EN ARCHIVO
// ============================================
function guardarHashEnArchivo(password, hash) {
    const contenido = [
        '='.repeat(60),
        'IPUC LA FONDA v5.0 - Hash de Contraseña',
        '='.repeat(60),
        `Salt: ipuc5_salt_2026`,
        `Contraseña: ${password}`,
        `Hash: ${hash}`,
        '='.repeat(60),
        '',
        '⚠️  Guarda este archivo en un lugar seguro'
    ].join('\n');
    
    // Crear blob y descargar
    const blob = new Blob([contenido], { type: 'text/plain' });
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = 'hash_ipuc.txt';
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
    
    console.log('✅ Hash guardado en hash_ipuc.txt');
}

// ============================================
// INTERFAZ HTML (SI SE ABRE EN NAVEGADOR)
// ============================================
if (typeof window !== 'undefined' && typeof document !== 'undefined') {
    document.addEventListener('DOMContentLoaded', function() {
        // Solo mostrar si no hay otro contenido
        if (document.body.children.length === 0 || window.location.pathname.includes('generar_hash')) {
            document.body.innerHTML = `
                <div style="max-width:500px;margin:50px auto;padding:30px;font-family:'Segoe UI',sans-serif;background:#fff;border-radius:16px;box-shadow:0 8px 32px rgba(0,0,0,0.1);">
                    <h2 style="color:#1a237e;text-align:center;">🔐 IPUC LA FONDA v5.0</h2>
                    <h3 style="text-align:center;color:#666;">Generador de Hash de Contraseña</h3>
                    
                    <div style="margin:20px 0;">
                        <label style="font-weight:600;display:block;margin-bottom:5px;">Contraseña:</label>
                        <input type="text" id="passwordInput" value="S3QyFkrCkNdRQk4" style="width:100%;padding:10px;border:2px solid #e0e0e0;border-radius:8px;font-size:16px;">
                    </div>
                    
                    <div style="margin:20px 0;">
                        <label style="font-weight:600;display:block;margin-bottom:5px;">Salt:</label>
                        <input type="text" id="saltInput" value="ipuc5_salt_2026" readonly style="width:100%;padding:10px;border:2px solid #e0e0e0;border-radius:8px;font-size:16px;background:#f5f5f5;">
                    </div>
                    
                    <button onclick="generarHashUI()" style="width:100%;padding:12px;background:#1a237e;color:white;border:none;border-radius:8px;font-size:16px;font-weight:600;cursor:pointer;">
                        🔒 Generar Hash
                    </button>
                    
                    <div id="resultado" style="margin-top:20px;padding:15px;background:#f5f5f5;border-radius:8px;display:none;">
                        <strong>Hash generado:</strong>
                        <p id="hashResultado" style="word-break:break-all;font-family:monospace;margin-top:8px;"></p>
                    </div>
                    
                    <button id="btnDescargar" onclick="descargarHash()" style="width:100%;padding:10px;margin-top:10px;background:#4caf50;color:white;border:none;border-radius:8px;font-size:14px;cursor:pointer;display:none;">
                        💾 Descargar hash.txt
                    </button>
                    
                    <p style="text-align:center;margin-top:20px;color:#666;font-size:14px;">
                        IPUC LA FONDA v5.0 | "Donde el Espíritu Santo se mueve"
                    </p>
                </div>
            `;
        }
    });
}

// ============================================
// FUNCIONES PARA LA INTERFAZ HTML
// ============================================
let ultimoHash = '';
let ultimaPassword = '';

function generarHashUI() {
    const password = document.getElementById('passwordInput').value;
    const salt = document.getElementById('saltInput').value;
    
    if (!password) {
        alert('Ingresa una contraseña');
        return;
    }
    
    const hash = generarHash(password, salt);
    
    document.getElementById('hashResultado').textContent = hash;
    document.getElementById('resultado').style.display = 'block';
    document.getElementById('btnDescargar').style.display = 'block';
    
    ultimoHash = hash;
    ultimaPassword = password;
    
    console.log('✅ Hash generado:', hash);
}

function descargarHash() {
    if (ultimoHash && ultimaPassword) {
        guardarHashEnArchivo(ultimaPassword, ultimoHash);
    }
}

// ============================================
// EXPORTAR FUNCIONES
// ============================================
window.generarHash = generarHash;
window.generarHashConsola = generarHashConsola;
window.guardarHashEnArchivo = guardarHashEnArchivo;
window.generarHashUI = generarHashUI;
window.descargarHash = descargarHash;

console.log('✅ IPUC LA FONDA - Generador de Hash v5.0 cargado');
console.log('💡 Ejecuta generarHashConsola() para ver el hash por defecto');
console.log('💡 O abre este archivo en el navegador para usar la interfaz gráfica');
